src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/FlowScannerTest.java

```diff
@@ -0,0 +1,492 @@
+        Predicate<FlowNode> outputPredicate = new Predicate<FlowNode>() {
+            @Override
+            public boolean apply(FlowNode input) {
+                if (input instanceof StepAtomNode) {
+                    StepAtomNode san = (StepAtomNode)input;
+                    StepDescriptor sd = san.getDescriptor();
+                    return sd != null && descriptorId.equals(sd.getId());
+                }
+                return false;
+            }
+        };
+        return outputPredicate;
+    }
+
+    Predicate<FlowNode> MATCH_ECHO_STEP = predicateMatchStepDescriptor("org.jenkinsci.plugins.workflow.steps.EchoStep");
+
+    static final class CollectingVisitor implements FlowNodeVisitor {
+        ArrayList<FlowNode> visited = new ArrayList<FlowNode>();
+
+        @Override
+        public boolean visit(@Nonnull FlowNode f) {
+            visited.add(f);
+            return true;
+        }
+
+        public void reset() {
+            this.visited.clear();
+        }
+
+        public ArrayList<FlowNode> getVisited() {
+            return visited;
+        }
+    };
+
+    /** Tests the basic scan algorithm, predicate use, start/stop nodes */
+    @Test
+    public void testSimpleScan() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "Convoluted");
+        job.setDefinition(new CpsFlowDefinition(
+                "sleep 2 \n" +
+                "echo 'donothing'\n" +
+                "echo 'doitagain'"
+        ));
+
+        /** Flow structure (ID - type)
+         2 - FlowStartNode
+         3 - SleepStep
+         4 - EchoStep
+         5 - EchoStep
+         6 - FlowEndNode
+         */
+
+        WorkflowRun b = r.assertBuildStatusSuccess(job.scheduleBuild2(0));
+        FlowExecution exec = b.getExecution();
+        AbstractFlowScanner[] scans = {new LinearScanner(),
+                new DepthFirstScanner(),
+                new ForkScanner()
+        };
+
+        List<FlowNode> heads = exec.getCurrentHeads();
+
+        // Iteration tests
+        for (AbstractFlowScanner scan : scans) {
+            System.out.println("Iteration test with scanner: "+scan.getClass());
+            scan.setup(heads, null);
+
+            for (int i=6; i>2; i--) {
+                Assert.assertTrue(scan.hasNext());
+                FlowNode f = scan.next();
+                Assert.assertEquals(Integer.toString(i), f.getId());
+            }
+
+            FlowNode f2 = scan.next();
+            Assert.assertFalse(scan.hasNext());
+            Assert.assertEquals("2", f2.getId());
+        }
+
+        // Block Hopping tests
+        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+        Assert.assertFalse("BlockHopping scanner jumps over the flow when started at end", scanner.setup(heads, Collections.EMPTY_SET));
+        List<FlowNode> collectedNodes = scanner.filteredNodes(Collections.singleton(exec.getNode("5")), null, (Predicate)Predicates.alwaysTrue());
```

Comment:
@amuniz Perhaps not as a filter (since you're not filtering without an API), but you can do the following if you really need this:  `scanner.filteredNodes(myHeadNode, Predicates.Predicates.alwaysTrue());`

I'm trying to avoid adding an API for that though, because I do *not* want users to use these to just collect up all the FlowNodes without filtering - we provide an iterator interface if they need to walk the full graph, and also implement Iterable to allow using it in the for-each loops (a bit unusual, but it solves the problem in the best way). 

If you're *cough* looking for the start of all blocks, there is: https://github.com/jenkinsci/workflow-api-plugin/pull/2/files#diff-86ccfab3a45f361209746daa6064ae0fR85